contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -3073,6 +3073,585 @@ CREATE OR REPLACE FUNCTION sys.substring(string sys.NCHAR, i INTEGER, j INTEGER)
+    END IF;
+
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.btrim(string::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+-- Additional handling is added for TRIM function with 2 arguments, 
+-- hence only following two definitions are required.
+CREATE OR REPLACE FUNCTION sys.TRIM(string sys.NVARCHAR, characters sys.VARCHAR)
+RETURNS sys.NVARCHAR
+AS 
+$BODY$
+BEGIN
+    IF string IS NULL OR characters IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.btrim(string::text, characters::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.TRIM(string sys.VARCHAR, characters sys.VARCHAR)
+RETURNS sys.VARCHAR
+AS 
+$BODY$
+BEGIN
+    IF string IS NULL OR characters IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.btrim(string::text, characters::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+-- wrapper functions for LTRIM
+CREATE OR REPLACE FUNCTION sys.LTRIM(string ANYELEMENT)
+RETURNS sys.VARCHAR
+AS
+$BODY$
+DECLARE
+    string_arg_datatype regtype;
+BEGIN
+    string_arg_datatype := pg_typeof(string);
+
+    -- restricting arguments with invalid datatypes for ltrim function
+    IF string_arg_datatype = 'text'::regtype OR
+        string_arg_datatype = 'ntext'::regtype OR
+        string_arg_datatype = 'sys.image'::regtype OR
+        string_arg_datatype = 'sys.sql_variant'::regtype OR
+        string_arg_datatype = 'xml'::regtype OR
+        string_arg_datatype = 'geometry'::regtype OR
+        string_arg_datatype = 'geography'::regtype  THEN
+            RAISE 'Argument data type % is invalid for argument 1 of ltrim function.', string_arg_datatype;
+    END IF;
+
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.ltrim(string::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.LTRIM(string sys.BPCHAR)
+RETURNS sys.VARCHAR
+AS
+$BODY$
+BEGIN
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.ltrim(string::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.LTRIM(string sys.VARCHAR)
+RETURNS sys.VARCHAR
+AS
+$BODY$
+BEGIN
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.ltrim(string::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.LTRIM(string sys.NCHAR)
+RETURNS sys.NVARCHAR
+AS
+$BODY$
+BEGIN
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.ltrim(string::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.LTRIM(string sys.NVARCHAR)
+RETURNS sys.NVARCHAR
+AS
+$BODY$
+BEGIN
+    IF string IS NULL THEN
+        RETURN NULL;
+    END IF;
+
+    RETURN PG_CATALOG.ltrim(string::text);
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION sys.LTRIM(string TEXT)
```

Comment:
For input type `TEXT/NTEXT`, precedence of `PG_CATALOG.LTRIM(TEXT)` function is more than `sys.LTRIM(anyelement)`. Hence added explicit function definition of `sys.LTRIM` with `TEXT` and `NTEXT` argument datatype so that any function call with `TEXT/NTEXT` input will use this `sys.LTRIM` definition

Reply:
LTRIM function call from TSQL endpoint should only consider definitions in `sys` schema, right? Until we have specified the `PG_CATALOG` schema.

Reply:
`LTRIM` function call without namespace from TSQL endpoint will consider all the available definitions of `LTRIM` across all the namespace and best candidate will be chosen.